File: Hot100/42. 接雨水.go
package main

// 给定 n 个非负整数表示每个宽度为 1 的柱子的高度图，计算按此排列的柱子，下雨之后能接多少雨水。

func main() {

}

// 双指针
// 一列一列求，每一列所能接的雨水数量等于当前列高度和它左右最高列相对低的那列的差决定
func trap(height []int) int {
	if height == nil || len(height) == 0 {
		return 0
	}

	max := func(a, b int) int {
		if a >= b {
			return a
		}
		return b
	}

	l, r := 0, len(height)-1 // 左右指针
	lmax, rmax := 0, 0       // 左侧和右侧最高的列
	res := 0                 // 总雨水数

	for l < r {
		// 更新左右侧最大值
		lmax = max(height[l], lmax)
		rmax = max(height[r], rmax)

		if lmax <= rmax {
			// 若左侧比右侧小，则遍历左指针，因为不用担心右面会漏
			res += lmax - height[l]
			l++
		} else {
			// 若右侧比左侧小，则遍历右指针，因为不用担心左面会漏
			res += rmax - height[r]
			r--
		}
	}

	return res
}

File: C++/42.接雨水.cpp

// 42. 接雨水
//
// 给定 n 个非负整数表示每个宽度为 1 的柱子的高度图，计算按此排列的柱子，下雨之后能接多少雨水。

#include <algorithm>
#include <stack>
#include <vector>

class Solution {
public:
    // 1. 双指针 -- 类似 11
    int trap(std::vector<int>& height)
    {
        int lmax = 0, rmax = 0; // 左右最高的列
        int res = 0; // 盛水量

        for (int l = 0, r = height.size() - 1; l < r;) {
            // 更新左右最大值
            lmax = std::max(lmax, height[l]);
            rmax = std::max(rmax, height[r]);

            // 将当前列的盛水量加到总结果
            res += lmax <= rmax ? lmax - height[l++] : rmax - height[r--];
        }

        return res;
    }

    // 2. 单调栈 -- 类似 84、239、739、496、503
    int trap(std::vector<int>& height)
    {
        int res = 0; // 盛水量
        std::stack<int> stack; // 单调递减栈
        for (int i = 0; i < height.size(); i++) {
            // 若栈不为空且当前元素大于栈顶元素，依次将栈内大于当前元素的值出栈，计算它能接的雨水数量
            while (!stack.empty() && height[i] > height[stack.top()]) {
                // 获取当前栈顶元素，并出栈
                int top = height[stack.top()];
                stack.pop();

                // 若当前栈顶元素为栈内第一个元素，直接退出循环然后将当前元素入栈，因为左面没有比它高的元素，构不成凹形无法盛接雨水
                if (stack.empty()) {
                    break;
                }

                int left = height[stack.top()]; // 左侧高列
                int right = height[i]; // 右侧高列
                int width = i - stack.top() - 1; // 左右侧高列间距
                int height = std::min(left, right) - top; // 当前列的高度差

                // 将当前列的盛水量加到总结果
                res += width * height;
            }

            // 若栈为空或当前元素小于栈顶元素了，将当前元素入栈
            stack.push(i);
        }

        return res;
    }
};
